### PR TITLE
feat(manager/npm): Optimize npm dedupe option

### LINKED ISSUE #25462: Optimize NPM package manager with dedupe option
### Discussed in https://github.com/renovatebot/renovate/discussions/25455

<div type='discussions-op-text'>

<sup>Originally posted by **jelhan** October 26, 2023</sup>
### Tell us more.

NPM package manager with `npmDedupe` option is not very performant. I run into timeouts with my project. When looking into why it takes so much longer with deduplication, I noticed that the implementation could be improved.

Currently it does the following if `npmDedupe` is in `postUpdateOptions`:

1. Run `npm install` _without_ the `--package-lock-only` flag. This causes all packages to be downloaded and a `node_modules/` folder being created. Very time consuming. https://github.com/renovatebot/renovate/blob/3df10279cd3422c0ff815f37d8d6870105929462/lib/modules/manager/npm/post-update/npm.ts#L54-L63
2. Run `npm dedupe` as a second step. https://github.com/renovatebot/renovate/blob/3df10279cd3422c0ff815f37d8d6870105929462/lib/modules/manager/npm/post-update/npm.ts#L132-L136

Two optimizations are possible:

1. `npm install` can deduplicate the dependencies directly with the `--prefer-dedupe` flag. Instead of `npm install --no-audit && npm dedupe` we can do `npm install --prefer-dedupe`, which is much faster.
2. No need to download and install the packages for deduplication. `--prefer-dedupe` and `--package-lock-only` flags are working well together.

The performance difference is huge. `npm install --prefer-dedupe --package-lock-only --no-audit` is finished within 1 second on my test project while `npm install --no-audit && npm dedupe` takes minutes.

_Note: Converted from bug report in #25454 to this feature request._</div>

### PR DESCRIPTION
**Pull request by jelhan**
## Changes

`--prefer-dedupe` flag will be used if `postUpdatOptions` contains `npmDedupe` instead of running `npm dedupe` after `npm install`.

Additionally it performs a lockfile only update using the `--package-lock-only` flag instead of downloading and installing all packages.

This change significantly improves performance of NPM manager with `npmDedupe` option.

It drops support for NPM < 7. Per https://github.com/renovatebot/renovate/discussions/25455#discussioncomment-7400018 I assume that's okay.

## Context

- closes #25462

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — CHANGES_REQUESTED** [review]
This needs:
- Comment about npmDedupe compatibility in docs
- Wait for v38 to merge due to dropping support for npm <7 with deduple

**Comment by jelhan** [comment]
> Wait for v38 to merge due to dropping support for npm <7 with deduple

Is there an ETA for v38?

**Comment by rarkins** [comment]
Not currently. We could also consider either:
1. Adding a new option npmDedupeX which does the new functionality (breaking), or
2. Changing npmDeDupe as you've done, but leaving a npmDedupeLegacy with old functionality (non-breaking, but with workaround)

**Comment by jelhan** [comment]
> We could also consider either:
> 
>     1. Adding a new option npmDedupeX which does the new functionality (breaking), or
> 
>     2. Changing npmDeDupe as you've done, but leaving a npmDedupeLegacy with old functionality (non-breaking, but with workaround)

I like this idea of adding the new flow with a new option first. It would unblock this PR, reduce the impact in case this change has unforeseen effects and still allow us dropping the old functionality in next major (e.g. by making `npmDedupe` an alias of `npmPreferDedupe`).

`npmPreferDedupe` might be a good name as it matches the NPM flag used.

File: docs/usage/configuration-options.md
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2975]:
```suggestion
| `npmDedupe`                  | Run `npm install` with `--prefer-dedupe` option. Needs npm >= 7.                                                                                                        |
```

The `npm` project styles their name in lowercase. We can simplify a bit by saying `needs x`.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Review by viceice — COMMENTED** [review]
why not simply check npm constraint and use new dedupe if npm isn't constraint to `<7`?

if we drop npm  <7 support, we also drop older node version support.

**Comment by HonkingGoose** [comment]
> why not simply check npm constraint and use new dedupe if npm isn't constraint to `<7`?

What happens then for user who do _not_ set a `npm` constraint in their `package.json` file?

**Comment by viceice** [comment]
> > why not simply check npm constraint and use new dedupe if npm isn't constraint to `<7`?
> 
> What happens then for user who do _not_ set a `npm` constraint in their `package.json` file?

then we use latest npm version when not detected from lockfile version

File: lib/modules/manager/npm/post-update/__snapshots__/npm.spec.ts.snap
Comment (jelhan) [inline lib/modules/manager/npm/post-update/__snapshots__/npm.spec.ts.snap:3]: The `generate lock files` test was run with `npmDedupe` post update option. I refactored the `generate lock files` test to run with default options. Additionally I added two new tests for `npmDedupe` post update option: One running without NPM version constraint, one with NPM 6.

It is expected that the old `generates lock files` test equals the `deduplicates dependencies after installation with NPM <= 6` test.

I can create another small PR, which is only doing the refactoring to existing tests if you prefer that. We could merge that before and see here only the changes related to adding additonal logic for NPM >= 7. Up to you.

**Review by jelhan — COMMENTED** [review]
I updated the implementation to keep existing flow for NPM <= 6 as proposed by @viceice.

File: lib/modules/manager/npm/post-update/npm.spec.ts
Comment (rarkins) [inline lib/modules/manager/npm/post-update/npm.spec.ts:182]: Instead of this, write a deep expect statement here confirming how many commands were run, plus looking for prefer dedupe flag

Comment (jelhan) [reply lib/modules/manager/npm/post-update/npm.spec.ts:182]: Do I understand correctly that you prefer to not use snapshot test but instead assert against the content of `execSnapshot`? Something like this:

```js
expect(execSnapshots).toHaveLength(1);
expect(execSnapshots[0].cmd).toContain('--prefer-dedupe');
```

Should I refactor the `deduplicates dependencies after installation with NPM <= 6` as well? Or should we keep it snapshot-based to not touch tests for existing functionality to much in this PR?

Comment (rarkins) [reply lib/modules/manager/npm/post-update/npm.spec.ts:182]: Ideally leave existing tests alone to make it easier to tell that they haven't been broken. But for the new test you understand correctly what I was suggesting

Comment (jelhan) [reply lib/modules/manager/npm/post-update/npm.spec.ts]: I refactored the two new tests to not use snapshots. Please let me know if I should refactor the existing test `generates lock files` as well. The snapshot of that test is changing anyways significantly as the test was using `npmDedupe` post update option previously.

File: lib/modules/manager/npm/post-update/npm.spec.ts
Comment (rarkins) [inline lib/modules/manager/npm/post-update/npm.spec.ts:37]: Can these changes be reverted?

File: lib/modules/manager/npm/post-update/npm.ts
Comment (viceice) [inline lib/modules/manager/npm/post-update/npm.ts:55]: I think we need to use satisfies here, because constraints can be multiple ranges

File: lib/modules/manager/npm/post-update/npm.spec.ts
Comment (viceice) [inline lib/modules/manager/npm/post-update/npm.spec.ts:184]: prefer to use `toMatchObject`

File: lib/modules/manager/npm/post-update/npm.spec.ts
Comment (viceice) [inline lib/modules/manager/npm/post-update/npm.spec.ts:208]: same here

File: lib/modules/manager/npm/post-update/npm.spec.ts
Comment (jelhan) [reply lib/modules/manager/npm/post-update/npm.spec.ts:37]: Sure. But test results would change significantly even than unless we add a NPM version constraint to use NPM 6.

Current test suit misses a test running with default configuration. That's the root cause here. I can refactor the tests in a separate PR which we merge before of you prefer that.

File: docs/usage/configuration-options.md
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2975]:
```suggestion
| `npmDedupe`                  | Run `npm install` with `--prefer-dedupe` for npm >= 7 or `npm dedupe` after `package-lock.json` update for npm <= 6.                                       |
```

File: lib/modules/manager/npm/post-update/npm.spec.ts
Comment (HonkingGoose) [inline lib/modules/manager/npm/post-update/npm.spec.ts:163]:
```suggestion
  it('deduplicates dependencies on installation with npm >= 7', async () => {
```

File: lib/modules/manager/npm/post-update/npm.spec.ts
Comment (HonkingGoose) [inline lib/modules/manager/npm/post-update/npm.spec.ts:187]:
```suggestion
  it('deduplicates dependencies after installation with npm <= 6', async () => {
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/modules/manager/npm/post-update/npm.spec.ts
Comment (jelhan) [reply lib/modules/manager/npm/post-update/npm.spec.ts:184]: @rarkins asked me to refactor away from using snapshots. (https://github.com/renovatebot/renovate/pull/25466#discussion_r1375258361) Wouldn't `expect(execSnapshot).toMatchObject()` be the same as a snaphot inlined into the test file itself? It's unclear to me how new tests should look like to be honest. Is there any documentation or best practice examples available?

File: lib/modules/manager/npm/post-update/npm.ts
Comment (jelhan) [reply lib/modules/manager/npm/post-update/npm.ts:55]: Not sure if I understand. I have tested this code with many ranges and I think it is working as expected:

```js
semver.gtr('7.0.0', '>= 7');
// false
semver.gtr('7.0.0', '8.x || >= 9');
// false
semver.gtr('7.0.0', '6');
// true
semver.gtr('7.0.0', '>=6');
// false
semver.gtr('7.0.0', '< 5');
// true
semver.gtr('7.0.0', '<5 || > 8');
// false
semver.gtr('7.0.0', '< 7');
// true
semver.gtr('7.0.0', '> 4');
// false
semver.gtr('7.0.0', '*');
// false
```

Do I miss something?

File: lib/modules/manager/npm/post-update/npm.spec.ts
Comment (HonkingGoose) [reply lib/modules/manager/npm/post-update/npm.spec.ts:184]: The [Development best practices docs](https://github.com/renovatebot/renovate/blob/main/docs/development/best-practices.md) has a section about unit testing, and other tips.

File: lib/modules/manager/npm/post-update/npm.ts
Comment (jelhan) [reply lib/modules/manager/npm/post-update/npm.ts:55]: `semver.intersects()` with a version range of `>=7.0.0` works as well and provides same results as the other solution:

```js
semver.intersects('>=7.0.0', '>= 7');
// true
semver.intersects('>=7.0.0', '8.x || >= 9');
// true
semver.intersects('>=7.0.0', '6');
// false
semver.intersects('>=7.0.0', '>=6');
// true
semver.intersects('>=7.0.0', '< 5');
// false
semver.intersects('>=7.0.0', '<5 || > 8');
// true
semver.intersects('>=7.0.0', '< 7');
// false
semver.intersects('>=7.0.0', '> 4');
// true
semver.intersects('>=7.0.0', '*');
// true
```

Happy to use `semver.intersects('>=7.0.0', npmToolConstraint.constraint)` if you prefer that for readability.

```suggestion
      semver.intersects('>=7.0.0', npmToolConstraint.constraint);
```

Comment (rarkins) [reply lib/modules/manager/npm/post-update/npm.ts:55]: The difference would be if someone has a constraint like `>=7.1.0` or `^7.1.0 || ^8.0.0`, etc. So yes please go ahead

Comment (viceice) [reply lib/modules/manager/npm/post-update/npm.ts]: please use intersects

**Review by rarkins — APPROVED** [review]

**Comment by jelhan** [comment]
@rarkins Thanks for approving. I haven't found information in the docs if I'm expected to merge latest changes from `main` into the feature branch or if you will do that before merging. I only found [information in contributing docs](https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#do-not-force-push-to-your-pull-request-branch) that I should not rebase. Please let me know if you expect me to update the branch.

**Comment by rarkins** [comment]
No, you don't need to merge changes from main, and in fact it's best not to as it will reset the test approval for first time contributors. @viceice can approve/merge once he's happy.

**Review by viceice — APPROVED** [review]

**Pull request merged**
